src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/XUnit/XUnitTestRunnerRunAsyncIntegration.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Datadog.Trace.Ci;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.ExtensionMethods;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.XUnit
+{
+    /// <summary>
+    /// Xunit.Sdk.TestRunner`1.RunAsync calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        Assemblies = new[] { "xunit.execution.dotnet", "xunit.execution.desktop" },
+        Type = "Xunit.Sdk.TestRunner`1",
+        Method = "RunAsync",
+        ReturnTypeName = "System.Threading.Tasks.Task`1<Xunit.Sdk.RunSummary>",
+        ParametersTypesNames = new string[0],
+        MinimumVersion = "2.2.0",
+        MaximumVersion = "2.*.*",
+        IntegrationName = IntegrationName)]
+    public static class XUnitTestRunnerRunAsyncIntegration
+    {
+        private const string IntegrationName = "XUnit";
+
+        static XUnitTestRunnerRunAsyncIntegration()
+        {
+            // Preload environment variables.
+            CIEnvironmentValues.DecorateSpan(null);
+        }
+
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget>(TTarget instance)
+        {
+            if (!Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationName))
+            {
+                return CallTargetState.GetDefault();
+            }
+
+            TestRunnerStruct runnerInstance = instance.As<TestRunnerStruct>();
+
+            // Skip test support
+            if (runnerInstance.SkipReason is null)
+            {
+                return CallTargetState.GetDefault();
```

Comment:
It seems to be that if there is no skip reason (in other words, we are going to run this test case), then we exit out and never create a span. Is that correct? I thought maybe there would be a corresponding `OnMethodEnd` that might do the work in this scenario

Reply:
Although, now I notice that the new TestInvoker`1.RunAsync instrumentation is doing more work than in the previous instrumentation. Is that the reason for this? Sorry, I'm not too knowledgeable on the hierarchy of the XUnit method calls

Reply:
Basically I'm just getting the tests that are going to be skipped with this integration. For other I just do the bypass. Is based in the same condition as XUnit source code: https://github.com/xunit/xunit/blob/main/src/xunit.v3.core/Sdk/v3/Runners/TestRunner.cs#L195

The idea is that a test is skipped if the `FactAttribute` or `TheoryAttribute` has something in the Skip property, and in case the `Assert.Skip` is used the skip reason is always filled at least with a token: https://github.com/xunit/assert.xunit/blob/7e0022ae9661014e16bd42cc5ac343a743e1ee72/Sdk/Exceptions/SkipException.cs#L23 . The token value: https://github.com/xunit/assert.xunit/blob/b4f2de311f3b050d3cce6d6024dcdccfcfa2bbf8/Sdk/DynamicSkipToken.cs#L12

For normal tests we Instrument the XUnit the TestInvoker, called from: https://github.com/xunit/xunit/blob/main/src/xunit.v3.core/Sdk/v3/Runners/TestRunner.cs#L221

The reason that I don't use the same integration is because here: https://github.com/xunit/xunit/blob/main/src/xunit.v3.core/Sdk/v3/Runners/TestRunner.cs#L290 the source code clears all the exceptions produced by the test, so I'm not able to include the exceptions in the test span at the OnEndMethod callback.

Reply:
Thanks for the in-depth explanation, I really appreciate it 😄